vespalib/src/vespa/vespalib/util/invokeserviceimpl.cpp

```diff
@@ -0,0 +1,86 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "invokeserviceimpl.h"
+#include <cassert>
+
+namespace vespalib {
+
+InvokeServiceImpl::InvokeServiceImpl(duration napTime)
+    : _naptime(napTime),
+      _lock(),
+      _currId(0),
+      _closed(false),
+      _toWakeup(),
+      _thread()
+{
+}
+
+InvokeServiceImpl::~InvokeServiceImpl()
+{
+    {
+        std::lock_guard guard(_lock);
+        assert(_toWakeup.empty());
+        _closed = true;
+    }
+    if (_thread) {
+        _thread->join();
+    }
+}
+
+class InvokeServiceImpl::Registration : public IDestructorCallback {
+public:
+    Registration(InvokeServiceImpl * service, uint64_t id) noexcept
+        : _service(service),
+          _id(id)
+    { }
+    Registration(const Registration &) = delete;
+    Registration & operator=(const Registration &) = delete;
+    ~Registration() override{
+        _service->unregister(_id);
+    }
+private:
+    InvokeServiceImpl * _service;
+    uint64_t        _id;
+};
+
+std::unique_ptr<IDestructorCallback>
+InvokeServiceImpl::registerInvoke(VoidFunc func) {
+    std::lock_guard guard(_lock);
+    uint64_t id = _currId++;
+    _toWakeup.emplace_back(id, func);
```

Comment:
I think std::move on the function here is useful, to save time for stuff that contains shared pointers etc. (the move constructor is noexcept in C++20)

Reply:
Agree